### PR TITLE
Added in new blueprint for handling interruptions

### PR DESCRIPTION
**Pull request by Ktmi**
This is intended to resolve kytos-ng/maintenance#59. This provides a blueprint for a mechanism to handle interruptions such as maintenance windows.

### CODE REVIEW

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:107]: It's typical for GET to take the filter params on its url query args. For this case though, can we instead provide the filters in the request body, and ideally following closely the same shape format that `maitenance` uses and `topology.interruptions.start`? Something like:

```
 {
        "switches": [<switch id>],
        "interfaces": [<interface id>],
        "links": [<link id>],
 }
```

That way, pretty much the payload can be reused to schedule a maintenance or an interruption once a network engineer is happy with the interruption preview result impact. Also, when filtering/querying for an interruption shouldn't we provide a filter for a service provider and which of its services instead of an interruption type?

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:137]: To simplify the query and filtering code on each provider, since each switch and link can be expressed as a function of interfaces, can we standardize to take a list of `interface` ids? 

Also, we might need to also specify another parameter to check if it's trying to check the interruption of an interface in a primary path or in a backup/failover path (not every service will have a concept of backup/faiolver path but for mef_eline it does). In this approach the caller client, topology, would need to send two requests and then also compare if impacted services had any backups or not, consolidating the response. Wdyt or how were you considering to address this part?

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:182]: Exceptionally for MW, we need to confirm if an actual shutdown is needed or not as a side effect so we might need to have a way to parametrize this. Will network engineers actually expect that a interruption of a MW actually also set interfaces physically down too or not? If they do, we need to make it clear that they'll need to undo the shutdown if mid maintenance they also want to run tests.

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:163]: Same comment for liveness and link up, later on, but with a lower priority, it might be worth to standardize them to use both ``topology.interruptions.start`` and ``topology.interruptions.end`` too, not sure if it'll be doable in this version, but let's keep on our backlog radar at least.

**Review by viniarck — COMMENTED** [review]
Nicely done @Ktmi, great how this has evolved and emerged as interruptions. 

Most of it looks good to me, asked minor implementation details clarifications though. Also, raised some questions that needs confirmation from network engineers to also confirm.

**Comment by viniarck** [comment]
Another adjacent thing to consider, once a service has is primary path interrupted but then it eventually ends the interruption, is it expected that the service should be using its primary path if it's available again or it's OK to continue in a backup or failover path? 

Currently, mef_eline won't switchover to the primary once it's available. In a production deployment, that means that network engineers would have to redeploy the EVCs if they want to force trying to use the primary path again.

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst:137]: A potential solution is to separate paths from services. Services specify which paths they may use, and interruptions query for which paths they affect.

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst:182]: Starting an interruption should neither disable nor deactivate a device, it should only set the status and status reason. This allows NApps to ignore interruptions by checking the status reason. If we add in a new parameter to `mef_eline` which specifies interruption types to ignore, we can then create paths for testing while an interruption is occurring.

File: docs/blueprints/EP035.rst
Comment (viniarck) [reply docs/blueprints/EP035.rst:137]: Yes. Sounds good to me, the client could parametrize which path like `?path_type="primary|backup"`. Trying to standardize this makes a lot of sense, and then it could be requested concurrently. 

Good thing that `mef_eline` evcs has the paths very well defined, so querying for that nested document object shouldn't be that difficult.

File: docs/blueprints/EP035.rst
Comment (viniarck) [reply docs/blueprints/EP035.rst:182]: > Starting an interruption should neither disable nor deactivate a device, it should only set the status and status reason. This allows NApps to ignore interruptions by checking the status reason.

Agreed, this is for sure the way that maximizes composability and minimizes side effects and surprises and easier to maintain. However, the question still remains: will network engineers actually expect that a interruption of a MW actually also set interfaces physically down too or not? 

This is a business requirement that needed to be sorted out with @italovalcy and @jab1982, can you sort that out? If they don't expect for interfaces to be shutdown, then it facilitates, but if they do, then we'll need to find a balance if they'll perform the shutdown via topology API or if we'll need to cover additional side effects.

Comment (viniarck) [reply docs/blueprints/EP035.rst:205]: > I chatted with @italovalcy. According to him maintenance is just supposed to make it so devices cannot participate in any services. He says that setting a device down isn't really semantically correct, but would ultimately want to have custom statuses. I told him that status reason basically implements customized statuses with the ability to be ignored, and he liked that.

Excellent. Great that as it is without actual putting devices down is sufficient for the production use case. Let's mark this conversation as resolved then.

Comment (viniarck) [reply docs/blueprints/EP035.rst:205]: > @viniarck do you know if any NApps treat EnittyStatus.DOWN and EnittyStatus.DISABLED differently?

On our kytos-ng org, no. We mostly have either conditionals for `status == EntityStatus.UP` or `status != EntityStatus.UP`. However, `status` has been recently exposed on `as_dict()` for every `GenericEntity` instances. So, still, they carry that semantics for any of the disabling or deactivation that historically weren't associated with adding a reason (that has been around for years). 

> Its conceivable that we could replace status entirely with status reasons by making EntityStatus.UP an empty frozen set, making all previous usage of status == EntityStatus.UP still valid, though status is EntityStatus.UP may become invalid.

Good observation. You can create a low priority issue to capture this idea for a future discussion. As of now, I don't think it's worth pursing that route, mainly due to much refactoring and breaking APIs. Let's keep evolving with `status_reason`, when the time comes we can reassess this again, OK?

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:140]: This thread https://github.com/kytos-ng/kytos/pull/356#discussion_r1157486183 is still unanswered. 

Can we only have service providers to not to have to filter for switches and links but only `interfaces`? 

Also, how are we going to specify which type of path? `path_type="primary|backup` ?

**Comment by viniarck** [comment]
> Another adjacent thing to consider, once a service has its primary path interrupted but then it eventually ends the interruption, is it expected that the service should be using its primary path if it's available again or it's OK to continue in a backup or failover path?
> 
> Currently, `mef_eline` won't switchover to the primary once it's available. In a production deployment, that means that network engineers would have to redeploy the EVCs if they want to force trying to use the primary path again.

This thread is still unanswered.

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:155]: In the original mw blueprint, it was also mentioned a desire to also have information about which user is associated to that service. Do we need to provide a richer reponse model here?

`{"id": <id>, "name": <service_name>, "owner": <str|null>}`, and then in the future potentially "username" and "email"?

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:129]: This phrase has become obsolete, right? Since it's taking the input in the request body now

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:165]: Same comment as https://github.com/kytos-ng/kytos/pull/356/files#r1163253185

File: docs/blueprints/EP035.rst
Comment (viniarck) [reply docs/blueprints/EP035.rst:140]: To avoid confusion what primary|backup means, it's worth noting that for mef_eline has two major cases:

- Dynamic circuits: their current "primary" is always `"current_path"`, and the `"failover_path"` is what is considered a "backup"
- Static circuits: `mef_eline` uses `"primary_path"` and `"backup_path"`, while the `"current_path"` will follow whichever path is active.

Comment (Ktmi) [reply docs/blueprints/EP035.rst:162]: > This thread [#356 (comment)](https://github.com/kytos-ng/kytos/pull/356#discussion_r1157486183) is still unanswered.

So for `mef_eline`, the idea would be that rather than querying `/v2/evc` for the appropriate paths, we instead query a new endpoint `/v3/paths` which is an aggregation of all paths for all evcs.

> Also, how are we going to specify which type of path? `path_type="primary|backup` ?

As for how to differentiate between primary and secondary, just include it in the name of the new path. I don't think it would be necessary to have to make two separate queries when we know we are going to ask for them at the same time.

> Can we only have service providers to not to have to filter for switches and links but only `interfaces`?

Maybe? But, what's your logic behind that idea?

Comment (viniarck) [reply docs/blueprints/EP035.rst:162]: > So for mef_eline, the idea would be that rather than querying /v2/evc for the appropriate paths, we instead query a new endpoint /v3/paths which is an aggregation of all paths for all evcs.

@Ktmi I had an impression that [on this thread](https://github.com/kytos-ng/kytos/pull/356#discussion_r1159884980), the client (`topology`) could perform two concurrent `GET <url for previewing>` for each `path_type`, that it'd be responsible for consolidating which service would be impacted based on their backup too. But for sure, that can also be delegated to each service provider to compute all of that too. Since you're considering this new `/v3/paths` can you give an example how it'd be potentially used including with payload inputs and response and `GET <url for previewing>` too, just so I can follow a full example?

Comment (Ktmi) [reply docs/blueprints/EP035.rst:162]: > > So for mef_eline, the idea would be that rather than querying /v2/evc for the appropriate paths, we instead query a new endpoint /v3/paths which is an aggregation of all paths for all evcs.
> 
> @Ktmi I had an impression that [on this thread](https://github.com/kytos-ng/kytos/pull/356#discussion_r1159884980), the client (`topology`) could perform two concurrent `GET <url for previewing>` for each `path_type`, that it'd be responsible for consolidating which service would be impacted based on their backup too. But for sure, that can also be delegated to each service provider to compute all of that too. Since you're considering this new `/v3/paths` can you give an example how it'd be potentially used including with payload inputs and response and `GET <url for previewing>` too, just so I can follow a full example?

This is what I'm thinking so far.

Inputs(Optional):

```json
{
    'type': <str>,
    'switches': [<str>],
    'interfaces': [<str>],
    'links': [<str>],
}
```

Expected output:

```json
[
    {
        'name': <evc_name>('(primary)'|'(backup)'|'(failover)'),
        'owner': <str or null>,
        'dynamic': <bool>,
        'path': <list[Links]>,
        'active': <bool>,
        'enabled': <bool>,
    },
]
```
Also note that active in this context should mean that the path is the current one in use by the EVC, and that enabled means that it is not being blocked by any interruptions. Now should we only care about enabled paths, active paths, or all paths when checking the affects of an interruption? Or do we extend `status_reason` into paths to check which interruptions they are already affected by, and filter out paths that are already affected by the interruption of the type we are checking for?

Comment (viniarck) [reply docs/blueprints/EP035.rst:162]: > Or do we extend status_reason into paths to check which interruptions they are already affected by

@Ktmi, I think it makes a lot of sense to extend `status_reason` into paths too. That way, we still have the flexibility of exposing the resulting `status_reason`, and if there are multiple ones that way it could be nicely exposed too. 

> and filter out paths that are already affected by the interruption of the type we are checking for?
 
I think not filtering them out would make it easier to understand. It'd be nice to have the `status` and `status_reason` of each path in the output too, it could make interpretation easier. At the end of the day an user could be trying to preview an interruption, so if it's already being interrupted, then makes sense to return the full consolidated `status_reason`, right? (related to the discussion of the first paragraph of this comment)

> Now should we only care about enabled paths, active paths, or all paths when checking the affects of an interruption?

If we're discussing here about `GET <url for previewing>`, since it's trying to preview an interruption, I think displaying all paths would be great. By all paths, I mean both primary and backup and ideally abstracted away from the semantics of the service provider (if an EVC is dynamic or static, regardless of the name of the path it can have one that represents a primary and another backup or secondary, just so in the future when we support other NApp that provision a service we can follow the same pattern you're standardizing here for interruption preview). I think your idea to aggregate by paths sounds really great to help with this but with path aggregation on it too if possible, for instance, on `GET <url for previewing>` something along these lines (and maybe also including the path links if you think it's useful):

```
[{
  "id": <id>,
  "name": < evc_name >,
  "owner": < str or null > ,
  "paths": {
    "primary": {
      "status": <status>
      "status_reason": <status_reason>
    },
    "backup": {
      "status": status
      "status_reason": <status_reason>
    }
  }
}]
```

That way on `GET /topology/v3/preview_interruption`, it could optionally also further expose the paths with the reasons too. Wdyt? 

Regarding `/v3/paths` would we still need it? I have an impression the way you're proposing and modelling `GET /topology/v3/preview_interruption` it could internally uses the collection controller accordingly with new aggregations without a new endpoint. Let me know if I misunderstood something. 

Regarding, `'active': <bool>, 'enabled': <bool>,` I have an impression that'd be a bit redundant, since if an EVC isn't active or enabled it can't be further interrupted, can we remove it from the output or do you see any other case where it can be useful?

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst]: From what I understand it would be nice, but not required. I've made including such information optional for now, just requiring at least an id.

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst:207]: We wouldn't need the new paths. Ultimately having paths separate opens up the possibility of testing a more distant scenario, where services can depend on other services, possibly across NApp boundaries, and interruptions will propagate recursively to help manage.

**Comment by Ktmi** [comment]
> > Another adjacent thing to consider, once a service has its primary path interrupted but then it eventually ends the interruption, is it expected that the service should be using its primary path if it's available again or it's OK to continue in a backup or failover path?
> > Currently, `mef_eline` won't switchover to the primary once it's available. In a production deployment, that means that network engineers would have to redeploy the EVCs if they want to force trying to use the primary path again.
> 
> This thread is still unanswered.

I spoke with @italovalcy and the response is that for dynamic paths, when the original interruption that affected them ends, continue to use the recomputed path, unless the recomputed path was using flexible metrics and didn't satisfy all criteria. For static paths, we switch back to the original path.

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:190]: There might be cases where NApps might want to subscribe for `topology.interruptions.end`, but typically higher level NApps like that primarly care about links interruptions should only probably listening to `kytos/topology.[link_down|link_up]`, but `topology` should for sure be listening to `topology.interruptions.[start|end]`, that way topology can still check if `status` is UP to notify only once accordingly. Do you share the same understanding @Ktmi? That way also facilitating for `topology` to notify only once when it's actually status UP or DOWN, so for instance if one interruption has ended, but there's still another topology impacting a link then topology wouldn't even send `kytos/topology.link_up`

File: docs/blueprints/EP035.rst
Comment (viniarck) [reply docs/blueprints/EP035.rst:207]: Right let's go with `GET <url for previewing>` as it is then. If in the future, we also need to include the paths in the output shouldn't be that hard to augment the response, and that wouldn't break API.

File: docs/blueprints/EP035.rst
Comment (viniarck) [inline docs/blueprints/EP035.rst:231]: In the future we could add which `<service provider name>` to preview for, and dispatch async with `httpx`, currently we only have `mef_eline` so it's worth implementing the filter since in practice it won't make much of a different, but let's make sure that this uses `httpx.AsyncClient` to play well with coroutines if we add another srevice provider in the future, OK?

**Comment by viniarck** [comment]
> For static paths, we switch back to the original path.

Cool. Please, later on map this requirement as an issue on mef_eline that's only desired circuits with static paths, that way one day it can be prioritized accordingly.

File: docs/blueprints/EP035.rst
Comment (viniarck) [reply docs/blueprints/EP035.rst]: Right, this part on the blueprint we're good since at both id and owner got covered, so we wouldn't need more effort on `maintenance`. Did you also confirm this with @italovalcy? If that was the understanding we can close this thread, otherwise we might need to create a new issue on mef_eline to support `owner: str on POST /v2/evc` cc'ing @Ktmi

**Comment by viniarck** [comment]
@Ktmi there's also a conflict with [`docs/blueprint/EP035.rst`](https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP035.rst), can you bump this as EP037? I've also closed most of the threads that I opened, followed up with minor ones to confirm we're on the same page.

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst]: Yeah, that makes sense. Continue to generate link up and down events, based on the changes induced by interruption start and end events.

Comment (viniarck) [reply docs/blueprints/EP035.rst]: Right. Sounds good. Let's close this thread then.

File: docs/blueprints/EP035.rst
Comment (viniarck) [reply docs/blueprints/EP035.rst]: David, once you create task(s) for this you can close this thread.

Comment (viniarck) [reply docs/blueprints/EP035.rst]: It's also related to this thread here https://github.com/kytos-ng/kytos/pull/356#discussion_r1157499941, both aren't blocking for this PR. It's up to you @Ktmi to see if you think you have capacity to also land these ones or if it'll be for a future release.

**Comment by viniarck** [comment]
@Ktmi, nicely done consolidating this blueprint, it's getting very close to land, I'm going to approve once these minor threads are closed/resolved:

- https://github.com/kytos-ng/kytos/pull/356#discussion_r1191495126
- https://github.com/kytos-ng/kytos/pull/356#discussion_r1187878480
- Finally, I got reminded of that initial requirement of potentially having the ability to redeploy a static circuit over a path that under maintenance (or provision a new static circuit), I'm not sure if @italovalcy will make use of this feature on `2023.1`, but it's worth confirming if we'll need to augment `mef_eline` to start potentially differentiating when a path that is still operationally active but status_reason == {"maintenance"}. 

Thanks, David.

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst]: Got confirmation from @italovalcy. Have it remain optional for now, as its unlikely it will be used in the short term.

File: docs/blueprints/EP035.rst
Comment (Ktmi) [reply docs/blueprints/EP035.rst]: Created a few PRs for this.

Comment (viniarck) [reply docs/blueprints/EP035.rst]: Great. It's nice to have it on our radar. Let's close the thread.

**Review by viniarck — APPROVED** [review]
Nicely done consolidating this blueprint @Ktmi. 

If anything else on `mef_eline` regarding provisioning paths during a maintenance is needed you can create a new issue, but since that won't impact much this blueprint, since the blueprint ended up very well agnostically solving the problem with interruptions, I'll go ahead and already leave it approved. If nobody has any extra comments or reviews we can merge this PR by Monday.

**Review by italovalcy — APPROVED** [review]
LGTM. I have a few questions/comments that I discussed with David/Vinicius on slack (specially the need for `SERVICE_PROVIDERS` setting versus using callback functions on top of existing events enhanced with a `dry-run` parameter). However, overall, the blueprint is good to go!

**Comment by viniarck** [comment]
Let's ship it then. Congrats @Ktmi, and thanks @italovalcy for you feedback. 

If anything else minor is needed it'll be augmented in a next iteration.

**Pull request merged**
